<a href="https://colab.research.google.com/github/twlee-dev/lab/blob/main/Day05_Ch08_%ED%86%B5%EA%B3%84%EC%8B%A4%EC%8A%B5_%EC%9D%B4%ED%83%9D%EC%9B%90.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 사전 준비: 라이브러리 및 데이터 로드


In [1]:
# 필요 라이브러리 설치
!pip install statsmodels

# 코랩에서 한글 폰트 사용을 위한 설정
!sudo apt-get install -y fonts-nanum
!sudo fc-cache -fv
!rm ~/.cache/matplotlib -rf

# 런타임 다시 시작 안내
# 위 코드 실행 후 상단 메뉴에서 [런타임] > [런타임 다시 시작]을 눌러주세요.
# 런타임을 다시 시작해야 한글 폰트가 적용됩니다.

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following NEW packages will be installed:
  fonts-nanum
0 upgraded, 1 newly installed, 0 to remove and 35 not upgraded.
Need to get 10.3 MB of archives.
After this operation, 34.1 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 fonts-nanum all 20200506-1 [10.3 MB]
Fetched 10.3 MB in 2s (6,862 kB/s)
debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 78, <> line 1.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
dpkg-preconfigure: unable to re-open stdin: 
Selecting previously unselected package fonts-nanum.
(Reading database ... 126374 files and dire

In [1]:
# 런타임 다시 시작 후, 이 셀을 실행하여 라이브러리와 폰트를 로드합니다.
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.formula.api as sm

# 한글 폰트 설정
plt.rc('font', family='NanumGothic')
# plt.rc('font', family='AppleGothic')
plt.rcParams['axes.unicode_minus'] = False # 마이너스 기호 깨짐 방지

In [6]:
# 실습용 'tips' 데이터셋을 불러옵니다.
tips = sns.load_dataset('tips')
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

### 📝 Statsmodels 사용법

1.  **데이터 분리 (X, y)**: 분석할 데이터를 **독립변수 `X`**(설명하는 변수들, DataFrame)와 **종속변수 `y`**(알고싶은 결과, Series)로 직접 나눕니다.
2.  **데이터 가공**:
    - **더미 변수 생성**: `pd.get_dummies()` 함수를 사용해 문자(범주형) 데이터를 0과 1로 이루어진 숫자 데이터로 직접 변환합니다.
    - **상수항(Intercept) 추가**: **가장 중요한 단계입니다.** 회귀식의 y절편(`b`ใน `y=ax+b`)을 계산하기 위해 `sm.add_constant(X)`를 사용하여 X 데이터에 상수항 컬럼을 **수동으로 추가**해야 합니다.
3.  **모델 학습**: `sm.OLS(y, X)` 또는 `sm.Logit(y, X)` 함수에 준비된 `y`와 `X` 데이터를 전달하여 모델을 학습시킵니다.

이제 이 3단계를 따라 모든 문제를 해결해 봅시다.


---


## 문제 1 (난이도: 하): 다중회귀분석의 첫걸음

> **🎯 목표:** 여러 개의 원인(독립변수)을 동시에 고려하여 결과를 예측하는 **다중회귀분석**을 파이썬 배열 방식으로 구현합니다.


### 💡 핵심 개념:

단순회귀가 하나의 원인(`X`)으로 결과를 설명했다면, **다중회귀**는 **여러 개의 원인(`X1, X2, ...`)**을 동시에 고려하여 결과를 더 정교하게 설명하는 방법입니다. 이때 각 원인의 영향력(계수)은 **'다른 원인들이 모두 동일하다고 통제했을 때'**의 순수한 영향력을 의미합니다.


### 📌 수행 과제:

1. `tips` 데이터셋에서 `tip`을 종속변수(y)로, `total_bill`과 `size`를 독립변수(X)로 분리하세요.
2. 독립변수(X)에 상수항을 추가하세요.
3. `sm.OLS`를 사용하여 다중회귀 모델을 학습시키고, `.summary()`로 결과를 확인하세요.
4. 결과표를 보고 `total_bill`과 `size`의 계수(coef)가 각각 무엇을 의미하는지 해석해 보세요.


### ✍️ 코드 작성:


In [9]:
# 1. 여기에 데이터를 y와 X로 분리하는 코드를 작성하세요.
# 종속변수 y에 'tip' 컬럼을 할당하세요.
import seaborn as sns
import statsmodels.api as sm
from statsmodels.api import add_constant
y = tips['tip']

# 독립변수 X에 'total_bill'과 'size' 컬럼을 리스트로 묶어 할당하세요.
tips = sns.load_dataset('tips')
X = tips[['total_bill', 'size']]

# 2. 여기에 상수항을 추가하는 코드를 작성하세요.
# statsmodels.api 모듈에서 add_constant 함수를 임포트하여 사용합니다.
from statsmodels.api import add_constant
X_const = add_constant(X)

# 3. 여기에 모델을 학습하고 결과를 출력하는 코드를 작성하세요.
# sm.OLS() 함수에 y와 X_const를 순서대로 넣어 모델을 만들고 학습시킵니다.
model_1 = sm.OLS(y, X_const).fit()
print(model_1.summary())

# 4. 결과 해석 (아래 주석에 직접 작성해 보세요)
# total_bill 계수의 의미: total_bill이 1단위 증가할 때, size가 동일하다면 tip은 해당 계수만큼 증가/감소합니다.
# size 계수의 의미: size가 1단위 증가할 때, total_bill이 동일하다면 tip은 해당 계수만큼 증가/감소합니다.

                            OLS Regression Results                            
Dep. Variable:                    tip   R-squared:                       0.468
Model:                            OLS   Adj. R-squared:                  0.463
Method:                 Least Squares   F-statistic:                     105.9
Date:                Thu, 11 Sep 2025   Prob (F-statistic):           9.67e-34
Time:                        02:52:06   Log-Likelihood:                -347.99
No. Observations:                 244   AIC:                             702.0
Df Residuals:                     241   BIC:                             712.5
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.6689      0.194      3.455      0.0

**결과 해석**

*   **total_bill 계수의 의미**: total_bill이 1단위 증가할 때, size가 동일하다면 tip은 해당 계수(0.0927)만큼 증가합니다.
*   **size 계수의 의미**: size가 1단위 증가할 때, total_bill이 동일하다면 tip은 해당 계수(0.1926)만큼 증가합니다.

두 변수 모두 p-value가 0.05보다 작으므로 통계적으로 유의미하게 팁에 영향을 미친다고 해석할 수 있습니다.

### 🤔 생각해 볼 문제:

`X_const.head()`를 출력해서 `sm.add_constant()`가 실제로 X 데이터에 어떤 변화를 주었는지 직접 눈으로 확인해보세요.


In [10]:
display(X_const.head())

const  total_bill  size
0    1.0       16.99     2
1    1.0       10.34     3
2    1.0       21.01     3
3    1.0       23.68     2
4    1.0       24.59     4

절편을 계산하면 어떤 결과가 나오나요?
회귀 모델에서 계산된 절편(intercept)은 model_1.summary() 결과표의 'const' 행에 있는 'coef' 값을 보면 알 수 있습니다. 문제 1의 결과표에서는 절편 값이 0.6689로 나타났습니다.

이 절편은 다른 모든 독립변수(total_bill, size)가 0일 때 종속변수(tip)의 평균적인 예상 값입니다. 하지만 이 데이터셋의 특성상 total_bill이나 size가 0인 경우는 현실적으로 없으므로, 이 절편 값 자체를 직접적으로 해석하는 것은 큰 의미가 없을 수 있습니다. 통계적으로는 다른 변수들의 영향력을 통제한 기준점 역할을 한다고 이해하시면 됩니다.



---


## 문제 2 (난이도: 하): 더미 변수 활용하기

> **🎯 목표:** '요일'과 같은 문자(범주형) 데이터를 분석에 활용하기 위해 **더미 변수**로 변환하는 방법을 배우고, 회귀분석에 적용합니다.


### 💡 핵심 개념:

**더미 변수**는 '성별', '요일'과 같은 문자(범주형) 데이터를 통계 모델에 사용하기 위해 **0과 1로 이루어진 숫자 변수**로 변환하는 기법입니다. Pandas의 `pd.get_dummies()` 함수를 사용하면 이 과정을 쉽게 처리할 수 있습니다. 이때, 여러 범주 중 하나는 **기준(reference)**이 되어 분석 오류를 막고 해석의 기준점 역할을 합니다.


### 📌 수행 과제:

1. `total_bill`과 '요일(`day`)'이 팁(`tip`)에 미치는 영향을 분석해봅시다.
2. `pd.get_dummies()`를 사용하여 'day' 컬럼을 더미 변수로 변환하세요. (단, `drop_first=True` 옵션 사용)
3. 기존 독립변수(`total_bill`)와 생성된 더미 변수를 합쳐 최종 X 데이터를 만드세요.
4. 상수항을 추가하고 회귀 모델을 학습시킨 후, 결과를 해석하세요.


### ✍️ 코드 작성:


In [13]:
# 1. 기본 데이터를 분리합니다.
import pandas as pd
import seaborn as sns
import statsmodels.api as sm

tips = sns.load_dataset('tips')
y = tips['tip']
X_base = tips[['total_bill']]

# 2. 여기에 'day' 컬럼으로 더미 변수를 생성하는 코드를 작성하세요.
# pd.get_dummies()를 사용하고, drop_first=True 옵션을 잊지 마세요.
day_dummies = pd.get_dummies(tips['day'], prefix='day', drop_first=True, dtype=int)

# 3. 여기에 기본 X 데이터와 더미 변수를 합치는 코드를 작성하세요.
# pd.concat을 사용하고, axis=1 옵션으로 옆으로 붙입니다.
X = pd.concat([X_base, day_dummies], axis=1)

# 4. 상수항을 추가하고 모델을 학습시킨 후 결과를 출력하세요.
X_const = sm.add_constant(X)
model_2 = sm.OLS(y, X_const).fit()
print(model_2.summary())

# 5. 결과 해석
# 기준이 된 요일은 무엇인가?:
# 'day_Sat' 계수의 의미:
#

                            OLS Regression Results                            
Dep. Variable:                    tip   R-squared:                       0.459
Model:                            OLS   Adj. R-squared:                  0.450
Method:                 Least Squares   F-statistic:                     50.67
Date:                Thu, 11 Sep 2025   Prob (F-statistic):           7.52e-31
Time:                        03:08:19   Log-Likelihood:                -350.03
No. Observations:                 244   AIC:                             710.1
Df Residuals:                     239   BIC:                             727.5
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.9205      0.186      4.943      0.0

**결과 해석**

*   **기준이 된 요일은 무엇인가?**: `drop_first=True` 옵션으로 인해 알파벳 순서로 가장 먼저 오는 요일인 'Thur'가 기준이 됩니다. (주석에 'Fri'로 잘못 기재되어 있었습니다.)
*   **'day_Fri' 계수의 의미**: `total_bill`이 동일할 때, 금요일(Fri)은 목요일(Thur) 대비 팁이 계수(0.0189)만큼 더 발생한다는 의미입니다. p-value(0.944)가 0.05보다 크므로 통계적으로 유의미하지 않습니다.
*   **'day_Sat' 계수의 의미**: `total_bill`이 동일할 때, 토요일(Sat)은 목요일(Thur) 대비 팁이 계수(-0.0671)만큼 덜 발생한다는 의미입니다. p-value(0.697)가 0.05보다 크므로 통계적으로 유의미하지 않습니다.
*   **'day_Sun' 계수의 의미**: `total_bill`이 동일할 때, 일요일(Sun)은 목요일(Thur) 대비 팁이 계수(0.0935)만큼 더 발생한다는 의미입니다. p-value(0.599)가 0.05보다 크므로 통계적으로 유의미하지 않습니다.

**참고**: 더미 변수의 p-value가 모두 유의미하지 않다는 것은 `total_bill`이 동일하다면 요일 자체는 팁 금액에 통계적으로 유의미한 영향을 미치지 않는다고 해석할 수 있습니다.

### 🤔 생각해 볼 문제:

`drop_first=True` 옵션을 빼고 `False`로 설정한 뒤 모델을 다시 학습시켜보세요. 결과가 어떻게 달라지나요? (힌트: `statsmodels`가 다중공선성에 대한 경고 메시지를 보여줄 수 있습니다.)


In [14]:
# drop_first=False로 설정하여 더미 변수 생성
# 문제 2의 코드를 기반으로 drop_first=True만 제거합니다.
import pandas as pd
import seaborn as sns
import statsmodels.api as sm

tips = sns.load_dataset('tips')
y = tips['tip']
X_base = tips[['total_bill']]

# day 컬럼으로 더미 변수 생성 (drop_first=False)
day_dummies_full = pd.get_dummies(tips['day'], prefix='day', dtype=int)

# 기본 X 데이터와 더미 변수 합치기
X_full = pd.concat([X_base, day_dummies_full], axis=1)

# 상수항 추가
X_full_const = sm.add_constant(X_full)

# 모델 학습 및 결과 출력
model_2_full = sm.OLS(y, X_full_const).fit()
print(model_2_full.summary())

                            OLS Regression Results                            
Dep. Variable:                    tip   R-squared:                       0.459
Model:                            OLS   Adj. R-squared:                  0.450
Method:                 Least Squares   F-statistic:                     50.67
Date:                Thu, 11 Sep 2025   Prob (F-statistic):           7.52e-31
Time:                        03:11:38   Log-Likelihood:                -350.03
No. Observations:                 244   AIC:                             710.1
Df Residuals:                     239   BIC:                             727.5
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.7455      0.131      5.677      0.0

요일 더미가 4개 모두 들어가 있어서(Thu, Fri, Sat, Sun) 아래 경고처럼 **더미 변수 함정(dummy variable trap)**이 발생했습니다.

“the design matrix is singular… strong multicollinearity”
→ 상수항 + 모든 요일 더미를 동시에 넣어서 완전 다중공선성이 생김


다중공선성이 있어 계수의 정확한 해석이 왜곡될 수 있음을 전제로 간단히 읽으면:

total_bill: 0.1047 (p<0.001)
청구금액이 $1 증가할 때(요일 동일), 팁은 약 $0.105 증가하는 경향이 통계적으로 유의합니다.

요일 더미
더미 함정 때문에 기준이 명확하지 않지만, 수치 자체로 보면

day_Sun: 0.2686 (p=0.026) → 일요일은 기준 대비 팁이 약 $0.27 더 높다는 효과가 유의하게 보입니다.

day_Sat, day_Thur, day_Fri는 p값이 크므로 유의하지 않음(기준 대비 차이가 뚜렷하다고 보기 어려움).

설명력: R² = 0.459 (중간 정도).
청구금액과 요일만으로 팁 변동의 약 46%를 설명.

요약

현재 모델은 더미 변수 함정으로 경고가 발생했습니다.

drop_first=True(또는 한 요일 더미를 수동으로 제거)로 다시 적합하세요. 기준은 보통 Fri가 됩니다.

그 상태에서 해석하면, 청구금액 효과는 강하고 유의, **일요일 프리미엄(≈$0.27)**만 유의하게 보이는 패턴이 확인될 가능성이 큽니다.

In [16]:
#기준 요일 1개 제거
import pandas as pd
import seaborn as sns
import statsmodels.api as sm

tips = sns.load_dataset('tips')
y = tips['tip']
X_base = tips[['total_bill']]

# 요일 더미: 기준 1개 제거(drop_first=True) → 기준은 사전순 첫 범주(Fri)
day_dummies = pd.get_dummies(tips['day'], prefix='day', drop_first=True, dtype=int)  # day_Sat, day_Sun, day_Thur

X = pd.concat([X_base, day_dummies], axis=1)
X_const = sm.add_constant(X)
model_2 = sm.OLS(y, X_const).fit()
print(model_2.summary())

                            OLS Regression Results                            
Dep. Variable:                    tip   R-squared:                       0.459
Model:                            OLS   Adj. R-squared:                  0.450
Method:                 Least Squares   F-statistic:                     50.67
Date:                Thu, 11 Sep 2025   Prob (F-statistic):           7.52e-31
Time:                        03:14:21   Log-Likelihood:                -350.03
No. Observations:                 244   AIC:                             710.1
Df Residuals:                     239   BIC:                             727.5
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.9205      0.186      4.943      0.0

더미함정 없이 잘 적합

기준(Reference) 요일

day_Fri, day_Sat, day_Sun만 있고 day_Thur가 없으므로 기준은 ‘Thur(목)’ 입니다.

계수 해석

total_bill = 0.1047 (p < 0.001)
다른 조건 동일할 때, 청구금액이 $1 증가하면 팁은 평균 $0.105 증가하는 경향이 유의합니다.

day_Fri = 0.0189 (p = 0.944)
금요일은 목요일 대비 팁이 $0.019 높게 나타나지만, 유의하지 않음.

day_Sat = -0.0671 (p = 0.697)
토요일은 목요일 대비 팁이 $0.067 낮게 나타나지만, 유의하지 않음.
→ 즉, **“total_bill이 같다면 토요일 효과는 목요일 대비 유의한 차이가 없다”**는 의미입니다.

day_Sun = 0.0935 (p = 0.599)
일요일은 목요일 대비 팁이 $0.094 높게 보이나, 유의하지 않음.

Intercept = 0.9205
(이론상) 목요일에 total_bill=0일 때의 팁. 현실적 의미는 제한적입니다.

한 줄 요약

팁은 청구금액(total_bill)의 증가에 강하게 비례하고, 요일 효과는(목 대비 금/토/일) 통계적으로 유의하지 않다고 볼 수 있습니다.



---


## 문제 3 (난이도: 중): 상호작용 효과 분석하기

> **🎯 목표:** 한 변수의 효과가 다른 변수의 상황에 따라 달라지는 **상호작용 효과**를 직접 변수를 곱하여 만들고, 그 의미를 해석합니다.


### 💡 핵심 개념:

**상호작용**은 한 변수의 효과가 다른 변수의 수준에 따라 달라지는 현상을 말합니다. 예를 들어, "식사 금액이 팁에 미치는 영향은 흡연 여부에 따라 다르다"와 같은 가설을 검증할 때 사용됩니다. 배열 방식에서는 상호작용 항을 **두 변수의 곱**으로 직접 만들어서 모델에 추가합니다.


### 📌 수행 과제:

1. "식사 총액(`total_bill`)이 팁에 미치는 영향은 흡연자(`smoker`) 그룹과 비흡연자 그룹에서 다를 것이다"라는 가설을 검증해봅시다.
2. `smoker` 변수를 더미 변수(`smoker_Yes`)로 만드세요. ('Yes' = 1, 'No' = 0)
3. `total_bill`과 `smoker_Yes`를 곱하여 상호작용 항(`bill_x_smoker`)을 만드세요.
4. `total_bill`, `smoker_Yes`, `bill_x_smoker`를 모두 포함하여 회귀 모델을 학습시키고 결과를 해석하세요.


### ✍️ 코드 작성:


In [20]:
# 1. 기본 데이터를 준비합니다.
import pandas as pd
import seaborn as sns
import statsmodels.api as sm

tips = sns.load_dataset('tips')
y = tips['tip']
X = tips[['total_bill']].copy()

# 2. 여기에 'smoker' 더미 변수를 X에 추가하는 코드를 작성하세요.
X['smoker_Yes'] = pd.get_dummies(tips['smoker'], drop_first=True, dtype=int)

# 3. 여기에 상호작용 항을 직접 계산하여 X에 추가하는 코드를 작성하세요.
# 'total_bill' 컬럼과 'smoker_Yes' 컬럼을 곱합니다.
X['bill_x_smoker'] = X['total_bill'] * X['smoker_Yes']

# 4. 상수항을 추가하고 모델을 학습시킨 후 결과를 출력하세요.
X_const = sm.add_constant(X)
model_3 = sm.OLS(y, X_const).fit()
print(model_3.summary())

# 5. 결과 해석 (아래 주석에 직접 작성해 보세요)
# 상호작용 항(bill_x_smoker)의 p-value는 유의미한가?:
# 상호작용 항 계수의 의미:

                            OLS Regression Results                            
Dep. Variable:                    tip   R-squared:                       0.506
Model:                            OLS   Adj. R-squared:                  0.500
Method:                 Least Squares   F-statistic:                     81.95
Date:                Thu, 11 Sep 2025   Prob (F-statistic):           1.56e-36
Time:                        03:22:57   Log-Likelihood:                -338.91
No. Observations:                 244   AIC:                             685.8
Df Residuals:                     240   BIC:                             699.8
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const             1.5643      0.238      6.570

상호작용 항(bill_x_smoker)의 p-value는 유의미한가?
네. p-value = 0.000 (유의수준 0.05보다 훨씬 작음) → 유의함.
⇒ 총액이 팁에 미치는 기울기(효과)가 흡연자와 비흡연자 사이에서 다릅니다.

상호작용 항 계수의 의미:
bill_x_smoker = 0.0676 → 흡연자의 기울기 − 비흡연자의 기울기 = 0.0676.
즉, 총액 $1 증가 시 흡연자 그룹의 팁 증가량이 비흡연자보다 $0.0676 더 큽니다.

### 🤔 생각해 볼 문제:

위 모델에서 비흡연자(smoker_Yes=0)의 `total_bill` 기울기는 얼마일까요? 그리고 흡연자(smoker_Yes=1)의 `total_bill` 기울기는 얼마일까요? (힌트: 흡연자의 기울기 = 기본 `total_bill` 기울기 + 상호작용 항 기울기)


In [21]:
# model_3의 회귀 계수를 가져옵니다.
coefs = model_3.params

# 비흡연자(smoker_Yes = 0)의 total_bill 기울기는 total_bill의 기본 계수입니다.
slope_nonsmoker = coefs['total_bill']

# 흡연자(smoker_Yes = 1)의 total_bill 기울기는 total_bill의 기본 계수와 상호작용 항 계수를 더한 값입니다.
slope_smoker = coefs['total_bill'] + coefs['bill_x_smoker']

print(f"비흡연자 그룹의 total_bill 기울기: {slope_nonsmoker:.4f}")
print(f"흡연자 그룹의 total_bill 기울기: {slope_smoker:.4f}")

비흡연자 그룹의 total_bill 기울기: 0.0696
흡연자 그룹의 total_bill 기울기: 0.1372


**결과 의미 해석**

*   **비흡연자 그룹의 total_bill 기울기 (0.0696)**: 비흡연자 그룹의 경우, 식사 총액(`total_bill`)이 1단위 증가할 때 팁(`tip`)은 평균적으로 약 0.0696 단위 증가하는 것으로 예상됩니다.
*   **흡연자 그룹의 total_bill 기울기 (0.1372)**: 흡연자 그룹의 경우, 식사 총액(`total_bill`)이 1단위 증가할 때 팁(`tip`)은 평균적으로 약 0.1372 단위 증가하는 것으로 예상됩니다.

상호작용 항 분석 결과와 마찬가지로, 흡연자 그룹의 `total_bill`에 대한 팁의 증가율(기울기)이 비흡연자 그룹보다 더 가파르다는 것을 확인할 수 있습니다. 이는 식사 총액이 증가할수록 흡연자 그룹이 비흡연자 그룹에 비해 더 많은 팁을 주는 경향이 있음을 시사합니다.

---


## 문제 4 (난이도: 중): 로지스틱 회귀분석

> **🎯 목표:** 결과가 '예/아니오' 같은 범주형일 때 사용하는 **로지스틱 회귀분석**을 구현하고, 그 결과를 **오즈비(Odds Ratio)**로 해석하는 방법을 배웁니다.


### 💡 핵심 개념:

**로지스틱 회귀분석**은 결과가 '예/아니오', '성공/실패'처럼 **두 가지 중 하나**일 때 사용하는 분석 방법입니다. 어떤 조건에서 특정 결과가 나타날 **'확률'**을 예측합니다. 회귀 계수는 그대로 해석하기 어렵고, `np.exp(계수)`를 계산한 **오즈비(Odds Ratio)**로 해석해야 합니다. 오즈비가 1보다 크면 확률 증가, 1보다 작으면 확률 감소를 의미합니다.


### 📌 수행 과제:

1. 식사 총액(`total_bill`)과 팁(`tip`) 금액을 보고, 그 고객이 '남성'일 확률을 예측하는 로지스틱 회귀 모델을 만들어봅시다.
2. 종속변수 `y`를 'sex'가 'Male'이면 1, 아니면 0이 되도록 변환하세요.
3. `sm.Logit` 함수를 사용하여 모델을 학습시키고, 결과를 확인하세요.
4. `np.exp()`를 이용해 계수를 오즈비로 변환하고, 그 의미를 해석하세요.


### ✍️ 코드 작성:


In [23]:
# 1. 여기에 종속변수 y를 0과 1로 변환하는 코드를 작성하세요.
# 'sex'가 'Male'이면 1, 아니면 0
y = tips['sex'].apply(lambda x: 1 if x == 'Male' else 0)

# 2. 독립변수 X를 준비하고 상수항을 추가하세요.
X = tips[['total_bill', 'tip']]
X_const = sm.add_constant(X)

# 3. 여기에 로지스틱 회귀 모델을 학습시키는 코드를 작성하세요.
# sm.Logit() 함수를 사용합니다.
model_4 = sm.Logit(y, X_const).fit()
print(model_4.summary())

# 4. 여기에 오즈비를 계산하고 출력하는 코드를 작성하세요.
import numpy as np
odds_ratios = np.exp(model_4.params)
print("\n--- 오즈비 (Odds Ratios) ---")
print(odds_ratios)

# 5. 결과 해석 (아래 주석에 직접 작성해 보세요)
# total_bill 오즈비의 의미: total_bill이 1단위 증가할 때, 남성일 오즈(odds)가 해당 오즈비만큼 배가 됩니다. (다른 변수 일정)
# tip 오즈비의 의미: tip이 1단위 증가할 때, 남성일 오즈(odds)가 해당 오즈비만큼 배가 됩니다. (다른 변수 일정)

Optimization terminated successfully.
         Current function value: 0.640304
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:                    sex   No. Observations:                  244
Model:                          Logit   Df Residuals:                      241
Method:                           MLE   Df Model:                            2
Date:                Thu, 11 Sep 2025   Pseudo R-squ.:                 0.01705
Time:                        03:26:08   Log-Likelihood:                -156.23
converged:                       True   LL-Null:                       -158.94
Covariance Type:            nonrobust   LLR p-value:                   0.06651
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0979      0.363     -0.270      0.787      -0.808       0.613
total_bill     0.0400      0.

**결과 해석**

로지스틱 회귀 모델 결과와 오즈비는 다음과 같이 해석할 수 있습니다.

*   **total_bill 오즈비 (1.040766)**: `total_bill`이 1단위 증가할 때, 고객이 남성일 오즈(odds)가 약 1.04배 증가합니다. 즉, 식사 총액이 많을수록 남성일 확률이 약간 높아지는 경향이 있습니다. 다만, `total_bill`의 P>|z| 값(0.073)이 0.05보다 크므로 통계적으로 유의미하다고 단정하기는 어렵습니다.

*   **tip 오즈비 (0.972066)**: `tip`이 1단위 증가할 때, 고객이 남성일 오즈(odds)가 약 0.97배가 됩니다. 즉, 팁 금액이 많을수록 남성일 확률이 약간 낮아지는 경향이 있으나, `tip`의 P>|z| 값(0.839)이 매우 크므로 통계적으로 유의미하지 않습니다.

결론적으로, 이 로지스틱 회귀 모델만으로는 `total_bill`과 `tip`이 고객의 성별을 예측하는 데 통계적으로 유의미한 영향을 미친다고 보기 어렵습니다.

### 🤔 생각해 볼 문제:

`model_4.predict(X_const)` 코드를 실행하면 각 손님에 대해 '남성일 확률'을 예측해줍니다. 이 예측 확률이 가장 높은 상위 5명의 손님 데이터(total_bill, tip)는 어떤 특징을 가지고 있는지 확인해보세요.


In [24]:
# model_4를 사용하여 남성일 확률을 예측합니다.
predicted_prob = model_4.predict(X_const)

# 예측 확률과 원래 데이터를 합쳐서 DataFrame을 만듭니다.
results_df = tips[['total_bill', 'tip']].copy()
results_df['predicted_male_prob'] = predicted_prob

# 예측 확률이 높은 순서대로 정렬하고 상위 5개를 선택합니다.
top_5_males = results_df.sort_values(by='predicted_male_prob', ascending=False).head(5)

# 상위 5명의 손님 데이터를 출력합니다.
display(top_5_males)

total_bill    tip  predicted_male_prob
156       48.17   5.00             0.843590
170       50.81  10.00             0.838758
59        48.27   6.73             0.837558
182       45.35   3.50             0.834105
102       44.30   2.50             0.832212

**결과 의미 해석**

`model_4.predict()`를 통해 예측된 '남성일 확률'이 가장 높은 상위 5명의 손님 데이터를 확인했습니다.

이 상위 5명의 데이터를 보면, 전반적으로 `total_bill` (식사 총액)이 높은 경향을 보입니다. 로지스틱 회귀 모델의 오즈비 해석에서 `total_bill`이 높을수록 남성일 오즈가 약간 증가하는 경향이 있다고 분석했는데, 이 결과는 해당 분석을 지지하는 것으로 볼 수 있습니다. `tip` 금액은 다양하게 나타나며, `total_bill`만큼 뚜렷한 패턴은 보이지 않습니다.

---


## 문제 5 (난이도: 중): 최고의 모델 찾기: AIC

> **🎯 목표:** 여러 모델 중 어떤 모델이 가장 좋은지 객관적인 점수로 평가하는 **AIC**의 개념을 이해하고, 이를 이용해 최적의 모델을 선택합니다.


### 💡 핵심 개념:

**AIC (Akaike Information Criterion)**는 여러 모델 중 어떤 모델이 가장 좋은지 평가하는 '점수'입니다. AIC는 **모델의 설명력(얼마나 잘 맞추는가)**은 높이고, **모델의 복잡성(얼마나 많은 변수를 썼는가)**에는 벌점을 줍니다. 따라서 **AIC 점수는 낮을수록 더 좋은(효율적인) 모델**이라고 평가합니다.


### 📌 수행 과제:

1. 팁(`tip`) 금액을 예측하기 위한 세 가지 다른 모델을 만듭니다.
   - 모델 A: `total_bill`만 사용
   - 모델 B: `total_bill`, `size` 사용
   - 모델 C: `total_bill`, `size`, `day`(더미) 사용
2. 각 모델의 `.aic` 속성을 확인하여 AIC 점수를 비교하세요.
3. 어떤 모델이 가장 좋은 모델인지 선택하고, 그 이유를 설명하세요.


### ✍️ 코드 작성:


In [26]:
import pandas as pd
import seaborn as sns
import statsmodels.api as sm

tips = sns.load_dataset('tips')
y = tips['tip']

# --- 모델 A: total_bill만 사용 ---
X_A = tips[['total_bill']]
X_A_const = sm.add_constant(X_A)
model_A = sm.OLS(y, X_A_const).fit()
print("Model A AIC:", model_A.aic)

# --- 모델 B: total_bill과 size 사용 ---
# 1. 여기에 모델 B의 X 데이터를 준비하고, 학습 후 AIC를 출력하는 코드를 작성하세요.
X_B = tips[['total_bill', 'size']]
X_B_const = sm.add_constant(X_B)
model_B = sm.OLS(y, X_B_const).fit()
print("Model B AIC:", model_B.aic)

# --- 모델 C: total_bill, size, day 사용 ---
# 2. 여기에 모델 C의 X 데이터를 준비하고, 학습 후 AIC를 출력하는 코드를 작성하세요.
# (힌트: 문제 2에서 X 데이터를 만들었던 과정을 참고하세요.)
day_dummies = pd.get_dummies(tips['day'], prefix='day', drop_first=True, dtype=int)
X_C = pd.concat([X_B, day_dummies], axis=1)
X_C_const = sm.add_constant(X_C)
model_C = sm.OLS(y, X_C_const).fit()
print("Model C AIC:", model_C.aic)

# 3. 결과 해석 (아래 주석에 직접 작성해 보세요)
# 가장 좋은 모델(AIC가 가장 낮은 모델)은?:
# 그렇게 생각하는 이유:

Model A AIC: 705.0761662637547
Model B AIC: 701.9702091086953
Model C AIC: 707.3838897060045


**결과 해석**

*   **가장 좋은 모델(AIC가 가장 낮은 모델)은?**: 모델 B
*   **그렇게 생각하는 이유**: AIC는 모델의 설명력과 복잡성을 동시에 고려하여 모델을 평가하는 지표이며, AIC 값이 낮을수록 더 좋은 모델로 간주됩니다. 모델 A의 AIC는 705.08, 모델 B의 AIC는 701.97, 모델 C의 AIC는 707.38입니다. 이 중에서 모델 B의 AIC 값이 가장 낮으므로, 세 모델 중에서는 모델 B가 가장 효율적인 모델이라고 평가할 수 있습니다.

### 🤔 생각해 볼 문제:

AIC는 모델의 복잡도(사용한 변수의 개수)에 벌점을 줍니다. 모델 B는 모델 A보다 변수가 하나 더 많음에도 AIC가 낮아졌습니다. 이는 무엇을 의미할까요?


AIC는 모델의 복잡성(변수 개수)에 벌점을 부여하지만, 동시에 모델이 데이터를 얼마나 잘 설명하는지(설명력)도 고려합니다.

모델 B가 모델 A보다 변수(size)가 하나 더 많음에도 불구하고 AIC가 더 낮아졌다는 것은, 새로 추가된 변수 size가 모델의 설명력을 충분히 향상시켜 변수 증가에 따른 페널티보다 설명력 향상 효과가 더 컸기 때문입니다. 즉, size 변수가 팁 예측에 유의미한 정보를 제공하여 모델의 효율성을 높였다고 해석할 수 있습니다.

## 오늘 실습한 문제의 주요 개념 종합정리 및 알아야 할 인사이트

### 1. 다중회귀분석 (문제 1)

*   **개념**: 여러 독립변수가 하나의 종속변수에 미치는 영향을 동시에 분석합니다. 각 변수의 계수는 다른 변수들을 통제했을 때의 순수한 영향력을 나타냅니다.
*   **Statsmodels 사용법**:
    *   종속변수(y)와 독립변수(X)를 분리합니다.
    *   `sm.add_constant(X)`를 사용하여 독립변수 데이터에 상수항 컬럼을 **수동으로 추가**해야 합니다. 이는 회귀식의 절편을 계산하는 데 필수적입니다.
    *   `sm.OLS(y, X_const).fit()`으로 모델을 학습시키고 `.summary()`로 결과를 확인합니다.
*   **인사이트**: 여러 요인이 결과에 복합적으로 영향을 미칠 때 각 요인의 상대적인 중요성과 방향(양의 영향인지 음의 영향인지)을 파악할 수 있습니다.

### 2. 더미 변수 활용 (문제 2)

*   **개념**: '요일', '성별' 등 범주형 데이터를 회귀 분석에 사용할 수 있도록 0과 1로 변환하는 기법입니다.
*   **Statsmodels 사용법**:
    *   `pd.get_dummies(데이터, prefix='접두사', drop_first=True)` 함수를 사용합니다.
    *   `drop_first=True` 옵션은 다중공선성 문제를 방지하기 위해 하나의 범주를 기준(reference)으로 삼아 해당 더미 변수를 제거합니다.
    *   생성된 더미 변수와 다른 독립변수를 `pd.concat()`으로 합쳐 최종 독립변수 데이터(X)를 만듭니다.
*   **인사이트**: 범주형 변수의 각 수준이 종속변수에 미치는 영향을 기준 범주와 비교하여 해석할 수 있습니다. `drop_first=True`를 통해 기준 범주를 명확히 설정하는 것이 중요합니다.

### 3. 상호작용 효과 분석 (문제 3)

*   **개념**: 한 독립변수가 종속변수에 미치는 영향이 다른 독립변수의 수준에 따라 달라지는 경우를 분석합니다. (예: 흡연 여부에 따라 식사 총액 대비 팁 증가율이 다름)
*   **Statsmodels 사용법**:
    *   상호작용 효과를 분석할 두 독립변수(하나는 더미 변수일 수 있음)를 **직접 곱하여** 새로운 상호작용 항 변수를 생성합니다.
    *   기존 독립변수들과 함께 상수항 및 상호작용 항을 모두 포함하여 회귀 모델을 학습시킵니다.
*   **인사이트**: 상호작용 항이 통계적으로 유의미하다면, 두 변수 간에 독립적인 영향 외에 추가적인 복합적인 관계가 존재함을 의미합니다. 상호작용 항의 계수는 한 변수의 효과가 다른 변수에 따라 얼마나 달라지는지를 나타냅니다.

### 4. 로지스틱 회귀분석 및 오즈비 (문제 4)

*   **개념**: 결과 변수가 '예/아니오', '성공/실패'와 같은 이진(binary) 범주형일 때, 각 조건에서 특정 결과가 나타날 **확률**을 예측하는 모델입니다.
*   **Statsmodels 사용법**:
    *   종속변수(y)를 0과 1로 인코딩해야 합니다.
    *   `sm.Logit(y, X_const).fit()`으로 모델을 학습시킵니다. (OLS 대신 Logit 사용)
    *   회귀 계수 자체보다는 `np.exp(계수)`를 계산한 **오즈비(Odds Ratio)**로 해석합니다.
*   **오즈비 해석**:
    *   오즈비 > 1: 독립변수가 1단위 증가할 때, 결과가 1일 오즈(odds)가 오즈비 배만큼 증가합니다.
    *   오즈비 < 1: 독립변수가 1단위 증가할 때, 결과가 1일 오즈가 오즈비 배만큼 감소합니다.
    *   오즈비 = 1: 독립변수의 변화가 결과의 오즈에 영향을 미치지 않습니다.
*   **인사이트**: 특정 조건이 발생할 확률이 다른 조건에 비해 얼마나 높은지(또는 낮은지)를 오즈비를 통해 정량적으로 파악할 수 있습니다.

### 5. AIC (Akaike Information Criterion) (문제 5)

*   **개념**: 여러 통계 모델의 성능을 비교하는 지표입니다. 모델의 **설명력**은 높게 평가하고 **복잡성(사용한 변수의 수)**에는 페널티를 부여합니다.
*   **해석**: **AIC 값이 낮을수록 더 좋은(데이터를 더 효율적으로 설명하는) 모델**입니다.
*   **Statsmodels 사용법**: 학습된 모델 객체의 `.aic` 속성을 통해 AIC 값을 쉽게 얻을 수 있습니다.
*   **인사이트**: AIC를 통해 모델의 설명력과 간결성 사이의 균형을 고려하여 최적의 모델을 선택하는 데 도움을 받을 수 있습니다. 단순히 설명력이 높다고 해서 (예: R-squared) 변수를 무작정 늘리는 것보다, AIC를 통해 모델의 효율성을 평가하는 것이 중요합니다.

이 개념들을 바탕으로 다양한 데이터를 회귀 분석하고 결과를 해석하는 능력을 키울 수 있습니다.